diff -rupN linux-2.6.34.orig/arch/arm/mach-omap2/board-rx51-peripherals.c linux-2.6.34/arch/arm/mach-omap2/board-rx51-peripherals.c
--- linux-2.6.34.orig/arch/arm/mach-omap2/board-rx51-peripherals.c	2010-06-01 23:00:19.615353659 +0930
+++ linux-2.6.34/arch/arm/mach-omap2/board-rx51-peripherals.c	2010-06-01 23:15:55.151351253 +0930
@@ -34,6 +34,8 @@
 #include <plat/onenand.h>
 #include <plat/gpmc-smc91x.h>
 
+#include <sound/tlv320aic3x.h>
+
 #include "mux.h"
 #include "hsmmc.h"
 
@@ -711,6 +713,10 @@ static struct twl4030_platform_data rx51
 	.vio			= &rx51_vio,
 };
 
+static struct aic3x_pdata rx51_aic3x_data __initdata = {
+	.gpio_reset		= 60,
+};
+
 static struct i2c_board_info __initdata rx51_peripherals_i2c_board_info_1[] = {
 	{
 		I2C_BOARD_INFO("twl5030", 0x48),
@@ -723,6 +729,7 @@ static struct i2c_board_info __initdata 
 static struct i2c_board_info __initdata rx51_peripherals_i2c_board_info_2[] = {
 	{
 		I2C_BOARD_INFO("tlv320aic3x", 0x18),
+		.platform_data = &rx51_aic3x_data,
 	},
 };
 
