diff -rupN linux-2.6.34.orig/sound/soc/codecs/tlv320aic3x.c linux-2.6.34/sound/soc/codecs/tlv320aic3x.c
--- linux-2.6.34.orig/sound/soc/codecs/tlv320aic3x.c	2010-06-01 22:43:44.763356923 +0930
+++ linux-2.6.34/sound/soc/codecs/tlv320aic3x.c	2010-06-01 22:45:23.835352194 +0930
@@ -998,7 +998,8 @@ static int aic3x_set_bias_level(struct s
 
 	switch (level) {
 	case SND_SOC_BIAS_ON:
-		/* all power is driven by DAPM system */
+		break;
+	case SND_SOC_BIAS_PREPARE:
 		if (aic3x->master) {
 			/* enable pll */
 			reg = aic3x_read_reg_cache(codec, AIC3X_PLL_PROGA_REG);
@@ -1006,13 +1007,8 @@ static int aic3x_set_bias_level(struct s
 				    reg | PLL_ENABLE);
 		}
 		break;
-	case SND_SOC_BIAS_PREPARE:
-		break;
 	case SND_SOC_BIAS_STANDBY:
-		/*
-		 * all power is driven by DAPM system,
-		 * so output power is safe if bypass was set
-		 */
+		/* fall through and disable pll */
 	case SND_SOC_BIAS_OFF:
 		if (aic3x->master) {
 			/* disable pll */
