diff -rupN linux-2.6.34.orig/arch/arm/mach-omap2/board-rx51-peripherals.c linux-2.6.34/arch/arm/mach-omap2/board-rx51-peripherals.c
--- linux-2.6.34.orig/arch/arm/mach-omap2/board-rx51-peripherals.c	2010-06-01 22:31:02.847350975 +0930
+++ linux-2.6.34/arch/arm/mach-omap2/board-rx51-peripherals.c	2010-06-01 22:50:35.635352586 +0930
@@ -418,9 +418,9 @@ static struct regulator_init_data rx51_v
 
 static struct regulator_init_data rx51_vmmc2 = {
 	.constraints = {
-		.name			= "VMMC2_30",
-		.min_uV			= 1850000,
-		.max_uV			= 3150000,
+		.name			= "V28_A",
+		.min_uV			= 2800000,
+		.max_uV			= 3000000,
 		.apply_uV		= true,
 		.valid_modes_mask	= REGULATOR_MODE_NORMAL
 					| REGULATOR_MODE_STANDBY,
